### ISSUE TITLE
Problems with training models with a single A100

### ISSUE THREAD

**Issue by yummyJacky** [YangRui2015/RiC#17]
Hi! Thanks for this amazing work and code open-sourcing!
I don't have enough experience in training large language models. Could you please give some suggestions on adjusting the hyperparameters when I use only an 80GB A100 to do the three-stage training of Ric. 

Question:I found that if I do not adjust the original hyperparameters, the loss after the training is basically unchanged or changes very little, and the obtained reward scores also change very little after the training had finished. It feels like the model training has not converged.

By the way, I notice that there is only one epoch in 'main.py'. Is it because multiple epochs are unnecessary or for some other reason?I have seen loss curves of other models before. The loss is unstable within an epoch, fluctuating up and down but basically unchanged. At the junction with the next epoch, the loss will suddenly drop.

Thanks in advance! 🙏

**Comment by yummyJacky** [comment]
Sorry,I noticed that the repo gives a note about the same training steps to reproduce the result.One thing I am  confused about is: Do four processes equal to four epochs?(Maybe a stupid question)

**Comment by YangRui2015** [comment]
Hi, I guess your issue may be due to using fewer steps than in the paper. If I remember correctly, using 4 processes with a batch size of 2 and 20,000 steps corresponds to about 2 epochs of the dataset. If you're using only one process, try increasing the batch size to 8 while keeping the same number of steps, or use a batch size of 4 with 40,000 steps. Let me know if this works for you!

**Comment by yummyJacky** [comment]
Thank you for your quick response!I will have the chance to try the experiment again this Sunday and reply here as soon as I get the result.
Thanks again!

**Comment by yummyJacky** [comment]
![Image](https://github.com/user-attachments/assets/e0fba200-68e4-4e76-8f6d-d6b1e6a72cfa)

![Image](https://github.com/user-attachments/assets/a0c37025-5a8b-4bab-a101-21c05c651807)

For the offline training, the hyperparameters I had changed are listed as below:
- batch_size=8(63GB/80GB)
- training_steps=80000(epoch=2.2)
- load_in_8bit=False and bf16=True(these two cannot be true at the same time and I found in this case the training speed is about 4x faster)
The result is shown in the Figure1. It seems that although the loss has decreased, it is not obvious.And at the junction of two epochs, the loss does not drop significantly as it is shown in Figure2 which I found in the other's SFT training experiment.
I don't know what's wrong in the experiment.Thank you for your suggestions!

**Comment by yummyJacky** [comment]
Maybe it is normal for the offline training because there are some more valuable data generated in the process of online training?Now the online training is going on and will finish tomorrow

**Comment by YangRui2015** [comment]
Your loss pattern differs from my experience. In my experience, the loss drops from nearly 2.0 to around 1.0 during offline training and continues decreasing to about 0.6 during online training, despite some fluctuations. I recommend setting load_in_8_bit=True, as we tuned the parameters for this configuration. Otherwise, you'll need to fine-tune the hyperparameters yourself, including LoRA rank and learning rate.

Once you observe a reasonable loss reduction, I recommend running the evaluation code instead of relying on the training loss.

**Comment by yummyJacky** [comment]
[training_metrics_offline_checkpoint20000_v2.pdf](https://github.com/user-attachments/files/19416005/training_metrics_offline_checkpoint20000_v2.pdf)

Thanks for your reply and suggestions!
There is also an experiment with the hyperparameters for about 19 hours with a single A100:
load_in_8bit=True and bf16=False
batch_size=8
training_steps=20000
However,the result is also not very optimistic.

**Comment by yummyJacky** [comment]
![Image](https://github.com/user-attachments/assets/201c3ff4-7443-48ba-b018-1bfbcced83d5)

Thanks for your reply and suggestions!
For comparison,there is also an experiment with the hyperparameters for about 19 hours with a single A100:
load_in_8bit=True and bf16=False（load_in_8bit=False and bf16=True about 4 hours training）
batch_size=8
training_steps=20000
However,the result is also not very optimistic.

**Comment by YangRui2015** [comment]
If so, one possible reason could be our latest update, as noted below:
`
we have updated the code to include the EOS token in data preparation for SFT and RiC, which helps resolve the issue of non-ending generation. As a result, the model can produce more natural responses, although the results may differ from the paper. If you wish to achieve the exact performance in the paper, please use an old version prior to November 3, 2024.
`

Another potential reason could be the Transformers library version. What version are you using?

**Comment by yummyJacky** [comment]
I just compared the versions again and all the package versions are consistent with those in requirements.txt
transformers==4.45.1

**Comment by yummyJacky** [comment]
I just compared the versions again and all the package versions are consistent with those in requirements.txt
transformers==4.45.1
I also calculated the mean and standard deviation of the data and compared it with the data mentioned in the previous issues.The below is about my data after prepareration:
[[-0.82138791  1.49611006]
 [ 1.29187156  1.74493306]]

**Comment by yummyJacky** [comment]
I don‘t know whether it is related to the loss but if I set the load_in_8bit True and bf16 False，I got the following warning which I didn't met when I set load_in_8bit False.

![Image](https://github.com/user-attachments/assets/a49ba8db-a535-4d5a-a8dc-a45725d850e9)

**Comment by YangRui2015** [comment]
The warning is fine. Could you try using an older version of the code, prior to November 3, 2024?

Additionally, you should first ensure that the vanilla SFT method is working before moving on to RiC. If vanilla SFT does not reduce loss, it is most likely not due to RiC-related issues. Let me know if vanilla SFT works.

**Comment by yummyJacky** [comment]
Thank you for your reply!
Actually the vanilla SFT method is also not working and I didn't notice or pay much attention to that,because the baseline method takes too long and I focus on Ric .This is the SFT result. I didn't change the hyperparameter so maybe it needs more steps to reach a reasonable loss value but now the loss decreases very little.

![Image](https://github.com/user-attachments/assets/58b52512-3b3a-4989-a608-2cf035a1ed7b)

**Comment by yummyJacky** [comment]
I will try the older version of the code according to the history commit today.

**Comment by YangRui2015** [comment]
It is wired. I didn't save the training loss for most experiments and mainly focus on the evaluation results. But I can find some preliminary results on wandb below. It's more reasonable.

![Image](https://github.com/user-attachments/assets/4a7f409c-7f40-47f2-b01f-c7df1ecb050f)

**Comment by yummyJacky** [comment]
Oh yeah!I just found a problem and will try again.As soon as I got the reasonable loss I will reply.Thank you very much

**Comment by yummyJacky** [comment]
Sorry,what I found in the morning didn't solve the problem and I have a new question maybe it is related to this.
When I do the data preparation there are two function build_dataset():
1.ric/utils.py
2.ric/prepare_dataset_with_rewards.py
And there are some differences especially when calculating the score,I'm not sure if it is related to the problem.When doing the data preparation I'm using the function build_dataset() in prepare_dataset_with_rewards.py

**Comment by yummyJacky** [comment]
Sorry,what I found in the morning didn't solve the problem and I have a new question maybe it is related to this.
When I do the data preparation there are two functions build_dataset():
1.ric/utils.py
![Image](https://github.com/user-attachments/assets/adc5e483-a4e8-4e31-ac8d-76403e14ec23)
2.ric/prepare_dataset_with_rewards.py
![Image](https://github.com/user-attachments/assets/659344aa-cd1b-40b4-ba66-7b3c6b1426fa)
And there are some differences especially when calculating the score,I'm not sure if it is related to the problem.When doing the data preparation I'm using the function build_dataset() in prepare_dataset_with_rewards.py

- the issue14:
![Image](https://github.com/user-attachments/assets/2f25dff0-95c2-4172-962d-a4ae3a3d0302)

But my data is a little different from this,mine is 
![Image](https://github.com/user-attachments/assets/7e2c4d87-21f3-4e10-9c76-cd0e227f782e)
Maybe it is because of different machines which I thought like this and do the offline training without thinking more

**Comment by yummyJacky** [comment]
> Sorry,what I found in the morning didn't solve the problem and I have a new question maybe it is related to this. When I do the data preparation there are two functions build_dataset(): 1.ric/utils.py ![Image](https://github.com/user-attachments/assets/adc5e483-a4e8-4e31-ac8d-76403e14ec23) 2.ric/prepare_dataset_with_rewards.py ![Image](https://github.com/user-attachments/assets/659344aa-cd1b-40b4-ba66-7b3c6b1426fa) And there are some differences especially when calculating the score,I'm not sure if it is related to the problem.When doing the data preparation I'm using the function build_dataset() in prepare_dataset_with_rewards.py

the function build_dataset() is never used in the experiment

**Comment by yummyJacky** [comment]
> Sorry,what I found in the morning didn't solve the problem and I have a new question maybe it is related to this. When I do the data preparation there are two functions build_dataset(): 1.ric/utils.py ![Image](https://github.com/user-attachments/assets/adc5e483-a4e8-4e31-ac8d-76403e14ec23) 2.ric/prepare_dataset_with_rewards.py ![Image](https://github.com/user-attachments/assets/659344aa-cd1b-40b4-ba66-7b3c6b1426fa) And there are some differences especially when calculating the score,I'm not sure if it is related to the problem.When doing the data preparation I'm using the function build_dataset() in prepare_dataset_with_rewards.py

![Image](https://github.com/user-attachments/assets/2f25dff0-95c2-4172-962d-a4ae3a3d0302)
Besides,my data is a little different from this,mine is 

![Image](https://github.com/user-attachments/assets/7e2c4d87-21f3-4e10-9c76-cd0e227f782e)

**Comment by YangRui2015** [comment]
I don’t think this issue is directly related to the loss. It might affect your final Pareto frontier, but it’s not the reason the loss isn’t degrading.

**Comment by yummyJacky** [comment]
Thank you very much
Could you please talk about the training loss pattern? like 1 or 2?Because I find that during the SFT sometimes the loss pattern is like 1

![Image](https://github.com/user-attachments/assets/2d50b1ef-dc4d-48f3-9e3e-41e3360a16a8)

![Image](https://github.com/user-attachments/assets/9d93d854-fe0e-4d26-ad28-398d1d19aae6)

**Comment by YangRui2015** [comment]
Pattern 1 is very common when training LLMs for repeated epochs due to memorization. There are some papers studying this phenomenon https://arxiv.org/pdf/2305.13230.

**Comment by yummyJacky** [comment]
Wow,thank you!This is also confusing me in the past few days because once I met this loss pattern when doing offline training although it was not very obvious.
This is an amazing work and thank you very much for your patient answer.I think there is still something wrong with the hyperparameters in my experience.
I wish you all the best and good luck to you for publishing more papers in the top conferences

**Comment by yummyJacky** [comment]
![Image](https://github.com/user-attachments/assets/98471d0e-b547-43b2-9e10-0458bc8c687c)
learning_rate = 5.656e-4  # 1.414e-4*4->5.656e-4
lr_scheduler = "cosine"  # linear->cosine
batch_size=8 #2->8
load_in_8bit=False
bf16=True
After adjusting the hyperparameters, the training loss is more reasonable than before.
Thx again!The experiment is going on(lr_scheduler is linear now,120000 steps for about 3 epochs in order to generate the third stage of the loss) and I had also added a valDataset(last 10% of the whole dataset) after every 3000 steps although in the code "do not evaluate during training".I'm not sure now whether it will affect the result.Because I found that maybe there is no loss computed in the evaluation.py

**Comment by YangRui2015** [comment]
Sounds good! Just a reminder to avoid overfitting if you train for too many epochs.

**Comment by yummyJacky** [comment]
okey,thank you very much for your suggestions!
